### ISSUE TITLE
Is gitDirty kept up-to-date?

### ISSUE THREAD

**Issue by UnkindPartition** [acfoltzer/gitrev#7]
Hi Adam,

From my reading of the source code, it doesn't look like the code will be recompiled when the repo becomes dirty (like when an untracked file is added, or a tracked file is changed). Is that correct?

**Comment by Blaisorblade** [comment]
Since the question is unanswered, allow me to butt in with answers and citations 😃 

**TL;DR. gitDirty is so up-to-date that using it can be annoying—but that's tracked through `.git/index`, so maybe it might miss changes if you don't run `git status` or such.** I'd never install a binary compiled without it, but YMMV.

IIUC, that is implemented through `addDependentFile index` — last commit on that IIUC is 2f71b53df026940e1a0e9eb9aef38cc318a33309 (from #6).

Here's people noticing that the detection is so accurate that it can impede the workflow (though that's before #6). The example uses `git add` though, explicitly touching the index).
https://github.com/commercialhaskell/stack/issues/1425

**Comment by UnkindPartition** [comment]
Are we talking about the same thing? By "dirty", I mean changes that you've made to your files in the working tree but haven't committed or staged, so they wouldn't be in the index.

**Comment by Blaisorblade** [comment]
You're right and I must (partially) retract—I expected git status to alter the index timestamp but that's just not true. OTOH, `git status` _always_ alters the timestamp of `.git` (by [creating and removing `index.lock`](http://stackoverflow.com/q/26937867/53974)).

The scenario from the OP that is still supported is "untracked file is added".

Do you think it's possible to fix this for `gitDirty` using only `addDependentFile`? I don't see how. I can only think of either always recompiling or adding additional ugly steps pre-build, via external tooling or _maybe_ Setup.hs (wild guess, don't know its API well enough).

**Comment by acfoltzer** [comment]
Sorry for the delay on this question; an email filter gone awry hid messages from this repo for months.

Indeed it is the case that the dirt detection is overly-sensitive, but as @Blaisorblade has ably laid out, it's not clear that there are any better alternatives. The fix in #6 made the problem better by only depending on `index` when actually asking for `isDirty`, so you don't incur the cost just by getting a hash.

My recommendation in the meantime is to try using `isDirty` in "front-endy" modules where possible, so that the overly-aggressive rebuilds don't hamper your development too much.
